cli/src/cmd/forge/geiger/find.rs

```diff
@@ -0,0 +1,254 @@
+//! scan a file for unsafe usage
+
+use crate::cmd::forge::geiger::error::ScanFileError;
+use forge_fmt::{parse, Visitable, Visitor};
+use foundry_common::fs;
+use solang_parser::{
+    diagnostics::Diagnostic,
+    pt::{ContractDefinition, Expression, FunctionDefinition, Loc, SourceUnit, Statement},
+};
+use std::{
+    convert::Infallible,
+    fmt,
+    path::{Path, PathBuf},
+};
+use yansi::Paint;
+
+/// Scan a single file for `unsafe` usage.
+pub fn find_cheatcodes_in_file(path: &Path) -> Result<SolFileMetrics, ScanFileError> {
+    let content = fs::read_to_string(path)?;
+    let cheatcodes = find_cheatcodes_in_string(&content)
+        .map_err(|diagnostic| ScanFileError::ParseSol(diagnostic, path.to_path_buf()))?;
+    Ok(SolFileMetrics { content, cheatcodes, file: path.to_path_buf() })
+}
+
+pub fn find_cheatcodes_in_string(src: &str) -> Result<CheatcodeCounter, Vec<Diagnostic>> {
+    let mut parsed = parse(src)?;
+    let mut visitor = CheatcodeVisitor::default();
+    let _ = parsed.pt.visit(&mut visitor);
+    Ok(visitor.cheatcodes)
+}
+
+#[derive(Default)]
+struct CheatcodeVisitor {
+    cheatcodes: CheatcodeCounter,
+}
+
+impl Visitor for CheatcodeVisitor {
+    type Error = Infallible;
+
+    fn visit_source_unit(&mut self, source_unit: &mut SourceUnit) -> Result<(), Self::Error> {
+        source_unit.0.visit(self)
+    }
+
+    fn visit_contract(&mut self, contract: &mut ContractDefinition) -> Result<(), Self::Error> {
+        contract.parts.visit(self)
+    }
+
+    fn visit_block(
+        &mut self,
+        _loc: Loc,
+        _unchecked: bool,
+        statements: &mut Vec<Statement>,
+    ) -> Result<(), Self::Error> {
+        statements.visit(self)
+    }
+
+    fn visit_expr(&mut self, _loc: Loc, expr: &mut Expression) -> Result<(), Self::Error> {
+        if let Expression::FunctionCall(loc, lhs, _) = expr {
```

Comment:
unfortunately, this does not cover all cases. consider the following tests
```rs
    #[test]
    fn can_find_nested_calls() {
        let s = r#"
        contract A is Test {
            function do_ffi() public {
                string[] memory inputs = new string[](1);
                abi.decode(vm.ffi(inputs), (address));
            }
        }
        "#;

        let count = find_cheatcodes_in_string(s).unwrap();
        assert_eq!(count.ffi.len(), 1);
        assert!(count.has_unsafe());
    }

    #[test]
    fn can_find_call_in_assignment() {
        let s = r#"
        contract A is Test {
            function do_ffi() public {
                string[] memory inputs = new string[](1);
                address stuff = vm.ffi(inputs);
            }
        }
        "#;

        let count = find_cheatcodes_in_string(s).unwrap();
        assert_eq!(count.ffi.len(), 1);
        assert!(count.has_unsafe());
    }
```

Reply:
both cases look correct to me, what am I missing?

Reply:
nested function calls, i think they would look like `FunctionCall(MemberAccess(FunctionCall(MemberAccess(...))))` and assignments `Expression::Assign`

Reply:
I see, so this is failing currently

```rust
 #[test]
    fn can_find_call_in_assignment() {
        let s = r#"
        contract A is Test {
            function do_ffi() public {
                string[] memory inputs = new string[](1);
                bytes stuff = vm.ffi(inputs);
            }
        }
        "#;

        let count = find_cheatcodes_in_string(s).unwrap();
        assert_eq!(count.ffi.len(), 1);
        assert!(count.has_unsafe());
    }
```
?